cpu/atxmega/include/periph_cpu.h

```diff
@@ -354,6 +354,213 @@ typedef enum {
+ * address have value equal to zero. To allow expand external RAM both
+ * @b RAM_LEN and @b EXP_RAM variables should be override at board
+ * makefile.include file.
+ *
+ * @note To avoid parser problems, @b RAM_LEN must be defined as decimal value.
+ *
+ * Example: Add 256K of external RAM
+ *
+ * The max addressable RAM by SP is 64K due to limit of 16 bits.  In this case,
+ * RAM will be 64K.  The remaining RAM can be addressed only by ebi_mem methods
+ * and GCC doesn't see it.
+ *
+ * At board/periph_conf.h:
+ *
+ * static const ebi_conf_t ebi_config = {
+ *   ...
+ *   .cs                     = {
+ *                              { EBI_CS_MODE_DISABLED_gc,
+ *                                0,
+ *                                EBI_CS_SRWS_0CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_DISABLED_gc,
+ *                                0,
+ *                                EBI_CS_SRWS_0CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_LPC_gc,
+ *                                EBI_CS_ASPACE_256KB_gc,
+ *                                EBI_CS_SRWS_1CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_DISABLED_gc,
+ *                                0,
+ *                                EBI_CS_SRWS_0CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                            },
+ * };
+ *
+ * At board/Makefile.include:
+ * override RAM_LEN = 65536
+ * override EXP_RAM = 1
+ *
+ * Example: Add 32K of external RAM and a LCD
+ *
+ * At board/periph_conf.h:
+ *
+ * static const ebi_conf_t ebi_config = {
+ * ...
+ *   .cs                     = {
+ *                              { EBI_CS_MODE_DISABLED_gc,
+ *                                0,
+ *                                EBI_CS_SRWS_0CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_DISABLED_gc,
+ *                                0,
+ *                                EBI_CS_SRWS_0CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_LPC_gc,
+ *                                EBI_CS_ASPACE_32KB_gc,
+ *                                EBI_CS_SRWS_1CLK_gc,
+ *                                0x0UL,
+ *                              },
+ *                              { EBI_CS_MODE_LPC_gc,
+ *                                EBI_CS_ASPACE_256B_gc,
+ *                                EBI_CS_SRWS_5CLK_gc,
+ *                                0x100000UL,
+ *                              },
+ *                            },
+ * };
+ *
+ * At board/Makefile.include:
+ * override RAM_LEN = 32768
+ * override EXP_RAM = 1
```

Comment:
>  * Example: Add 32K of external RAM and a LCD

I think you got your example wrong :wink: 
Isn't `RAM_LEN` the total RAM length? So it should be 32k + whatever the CPU has

Why not have 

```
EXT_RAM_LEN = 32768
```

and later in the CPU Makefile

```Make
EXT_RAM_LEN ?= 0

ifneq ($(EXT_RAM_LEN),0)
  RAM_LEN := $$(($(RAM_LEN) + $(EXT_RAM_LEN)))
  RAM_LEN := $$(($(RAM_LEN) > 65535 ? 65535 : $(RAM_LEN)))

  CFLAGS += -DRAM_LEN=$(RAM_LEN)-1
  LDSCRIPT_EXTRA = -Wl,--defsym=__heap_end=0x800000+$(RAM_LEN)-1
endif
```

Reply:
Hi @benpicco ,

Maybe docs aren't clear yet.

Tell me if below clarify to add on docs:

The memory address should be aligned at external memory size boundary, respecting power of 2. In this case, to add 32K memory on the system, the chip select address should be set to 0, or 32k, or 64 etc.

When board have external memory and user wants to map part of that memory to expand RAM, both internal and external memories must have same start address `A`. This is necessary to make sure RIOT-OS have a contiguous address space. The drawback is that first `Xk` external memory will be lost. 

Assuming internal top address memory is `B` and external top address memory is `C` . XMEGA will automatically access internal memory `<= B` when address collide with external memory. At address `>= B + 1`, XMEGA will access memory thru EBI.

```
 C      ------
        |    | 
        |    |
        |    |
 B -----|    |
   |    |    |
   |    |    |
 A |____|____|
```

> Isn't RAM_LEN the total RAM length? So it should be 32k + whatever the CPU has

Yes, it is the total length. It'll always match with external memory size and it should not be higher than 64kiB because SP is only 16 bit.

Reply:
Ah thank you for the clarification! I thought we could just map the external memory after the internal memory - better add this to the documentation then. Also that RAM_LEN must be a power of 2, so we can't e.g. have 48k RAM, right?

Reply:
> Also that RAM_LEN must be a power of 2, so we can't e.g. have 48k RAM, right?

Yes, this is based on a enum from CPU register which starts with 256B and goes up to 16MB.
I'll add above notes on docs.